### PR TITLE
Data release for manuscript

### PR DESCRIPTION
**Pull request by wdwatkins**
File path cleanup, readme, conda environment yml

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,18 @@
+#some of these may still be checked in; also using this
+#to ignore files when zipping repo for data release
+run_lstm_check_results.ipynb
```
Comment (aappling-usgs) [inline .gitignore]: not covered by `*.ipynb` below?

File: README.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+temperature = pd.read_csv('datarelease/temperature_observations.csv')
+forcing = pd.read_csv('datarelease/forcings.csv')
+
+forcing_temp = forcing.merge(temperature, how = 'outer')
+forcing_temp_renamed = forcing_temp.rename(columns={"discharge(cfs)" : "00060_Mean", "wtemp(C)" : "00010_Mean"})
+
+```
+### Explanation of different experiments
+
+At this point, you can decide which experiment to replicate.`experiments.csv` in the data release describes the various experiments run in this data release.  These columns are most important:
```
Comment (aappling-usgs) [inline README.md]:
```suggestion
At this point, you can decide which experiment to replicate. `experiments.csv` in the data release describes the various experiments run in this data release. These columns are most important:
```

File: README.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+- train_filter: Apply this filter to forcing data to filter out training and test data
+- test_filter: apply this filter to get *only* the test data. Test/train time period split is handled internally in the modeling code when running, use this to extract sites for a particular test set when checking performance.
+- hidden_sizes:  The sizes of hidden layers of the neural network.  Set these accordingly in StreamTemp-Integ.py
+- batch_size: Size of training batch.  Also set in StreamTemp-Integ.py
+
+To filter the sites and forcing data based on the first experiment in experiments.csv, and write the input files:
+```
+attr_filtered = attr.query('dag >= 60')
+attr_filtered
```
Comment (aappling-usgs) [inline README.md]: no need to print `attr_filtered` as part of the example code

File: README.md [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+attr_filtered.reset_index().to_feather('in/attr/attr_99.feather', version = 1)
+```
+
+Now you are almost ready to run the model code.
+
+## Run model
+
+First, set the forcing file name, site attributes name, batch size and hidden layer 
+sizes in StreamTemp-Integ.py on the line numbers shown.  Note that you only specify the base filename here,
+the directory (`in`) is specified elsewhere.
```
Comment (aappling-usgs) [inline README.md]:
```suggestion
the directories (`in/forcing` and `in/attr`) are specified elsewhere.
```

File: README.md [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+forcing_temp_renamed_filter.reset_index().to_feather('in/forcing/forcing_99.feather', version = 1)
+attr_filtered.reset_index().to_feather('in/attr/attr_99.feather', version = 1)
+```
+
+Now you are almost ready to run the model code.
+
+## Run model
+
+First, set the forcing file name, site attributes name, batch size and hidden layer 
+sizes in StreamTemp-Integ.py on the line numbers shown.  Note that you only specify the base filename here,
```
Comment (aappling-usgs) [inline README.md]:
```suggestion
sizes in StreamTemp-Integ.py on the line numbers shown.  Note that you only specify the base filename here;
```

File: README.md [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        print(files[i])
+        init_array.append(np.load(files[i])[:,:,0])
+    return(init_array)
+        
+init_array = load_initializations(initialization_preds)
+init_means = np.mean(init_array, axis = 0) 
+```
+
+
+#to compare to observations
```
Comment (aappling-usgs) [inline README.md]:
```suggestion
# Compare to observations
```

File: README.md [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+From a terminal:
+`python StreamTemp-Integ.py`
+
+Or from an iPython notebook:
+`%run StreamTemp-Integ.py`
+
+## Examine outputs
+Output files will be written to the `output` folder in a subdirectory named according to various model hyperparameters, the start/end dates of the training data, and the random number seed used.  `obs.npy` contains observations for the test time period, and `pred.npy` contains the test period predictions.  The model script is set up to run train the model for six different random number seeds (for size different random neural network weight initializations).
+
+#average together initializations
```
Comment (aappling-usgs) [inline README.md]:
```suggestion
# Compute ensemble predictions
```

File: README.md [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+Now you can train and make predictions with the model.
+
+From a terminal:
+`python StreamTemp-Integ.py`
+
+Or from an iPython notebook:
+`%run StreamTemp-Integ.py`
+
+## Examine outputs
+Output files will be written to the `output` folder in a subdirectory named according to various model hyperparameters, the start/end dates of the training data, and the random number seed used.  `obs.npy` contains observations for the test time period, and `pred.npy` contains the test period predictions.  The model script is set up to run train the model for six different random number seeds (for size different random neural network weight initializations).
```
Comment (aappling-usgs) [inline README.md:106]:
```suggestion
Output files will be written to the `output` folder in a subdirectory named according to various model hyperparameters, the start/end dates of the training data, and the random number seed used.  `obs.npy` contains observations for the test time period, and `pred.npy` contains the test period predictions.  The model script is set up to run train the model for six different random number seeds (and thus six different random neural network weight initializations).
```

File: README.md [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+```
+#load in observations, dropping degenerate third dimension
+obs = np.load('output/epochs2000_batch47_rho365_hiddensize100_Tstart20101001_Tend20141001_1/All-2010-2016/pred.npy')[:,:,0]
+#generate variety of metrics, or write own code as desired
+from hydroDL.post import plot, stat
+err = stat.statError(init_means, obs)
+
+```
+
+#to compare to specific test sites
```
Comment (aappling-usgs) [inline README.md]:
```suggestion
# Compare to specific test sites
```

File: README.md [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+
+Set `forcing_list` and `attr_list` (lines 19-20, seen above) to the new file names.
+ 
+Now compare the predictions and observations as before:
+```
+obs = np.load('TempDemo/FirstRun/epochs2000_batch50_rho365_hiddensize100_Tstart20101001_Tend20141001_8/All-2010-2016/obs.npy')[:,:,0]
+pred = np.load('TempDemo/FirstRun/epochs2000_batch50_rho365_hiddensize100_Tstart20101001_Tend20141001_8/All-2010-2016/pred.npy')[:,:,0]
+
+pred.shape
+obs.shape
```
Comment (aappling-usgs) [inline README.md:163]: You can delete these two lines

File: README.md [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+from hydroDL.post import plot, stat
+err = stat.statError(init_means, obs)
+
+```
+
+#to compare to specific test sites
+```
+#filter forcings to sites, based on test_filter column in experiments.csv
+attr_filtered_test = attr.query('dag == 60')
+attr_filtered
```
Comment (aappling-usgs) [inline README.md:146]:
```suggestion
```

File: README.md [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+        
+init_array = load_initializations(initialization_preds)
+init_means = np.mean(init_array, axis = 0) 
+```
+
+
+#to compare to observations
+```
+#load in observations, dropping degenerate third dimension
+obs = np.load('output/epochs2000_batch47_rho365_hiddensize100_Tstart20101001_Tend20141001_1/All-2010-2016/pred.npy')[:,:,0]
```
Comment (aappling-usgs) [inline README.md:134]: Seems weird that `obs` gets loaded from a file called `pred.npy`...what's happening here?

File: README.md [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+                       
+def load_initializations(files):
+    init_array = []
+    for i in range(len(files)):
+        print(files[i])
+        init_array.append(np.load(files[i])[:,:,0])
+    return(init_array)
+        
+init_array = load_initializations(initialization_preds)
+init_means = np.mean(init_array, axis = 0)
```
Comment (aappling-usgs) [inline README.md:127]: Can we call these `ensemble_means` instead? The 'initialization' naming is confusing to me because it sounds like these tables will contain initialization weights or perhaps the predictions from the just-initialized models before they undergo training (neither of which comes close to truth)

File: StreamTemp-Integ.py [hunk 14]
```diff
@@ -16,16 +16,12 @@
...
-attr_list = [
-             'attr_temp99%_days_99sites.feather'
-             ]
+forcing_list = ['forcing_99%_days_99sites_newv1.feather']
+attr_list = ['attr_temp99%_days_99sites_newv1.feather']
 
 Batch_list = [ 47]
 Hidden_list = [100, 100, 100]
-Randomseed = [ 1, 2, 3, 4, 5]
+Randomseed = [8,9,10,11,12,13]
```
Comment (aappling-usgs) [inline StreamTemp-Integ.py:24]: why this change?

File: StreamTemp-Integ.py [hunk 15]
```diff
@@ -48,7 +44,7 @@
         BATCH_SIZE = b_list
         RHO = 365
         HIDDENSIZE = h_list
-        saveEPOCH = 100   # it was 50
+        saveEPOCH = 200   # it was 50
```
Comment (aappling-usgs) [inline StreamTemp-Integ.py:47]: why this change?

File: StreamTemp-Integ.py [hunk 16]
```diff
@@ -68,13 +64,17 @@
...
         # Modify this based on your own location
-        rootDatabase = os.path.join(os.path.sep, absRoot, 'scratch', 'SNTemp')  # CAMELS dataset root directory: /scratch/Camels
-        rootOut = os.path.join(os.path.sep, absRoot, 'TempDemo', 'FirstRun')  # Model output root directory: /data/rnnStreamflow
-
-        forcing_path = os.path.join(os.path.sep, rootDatabase, 'Forcing', 'Forcing_new', f_list)  # obs_18basins
-        forcing_data =[]#pd.read_feather(forcing_path)
-        attr_path = os.path.join(os.path.sep, rootDatabase, 'Forcing', 'attr_new', a_list)
-        attr_data =[]#pd.read_feather(attr_path)
+        rootDatabase = os.path.join(os.path.sep, absRoot, 'in')  # CAMELS dataset root directory: /scratch/Camels
+        rootOut = os.path.join(os.path.sep, absRoot, 'output')  # Model output root directory: /data/rnnStreamflow
```
Comment (aappling-usgs) [inline StreamTemp-Integ.py:68]: These are nice directory simplifications in these surrounding lines. Getting nitpicky, but `in` and `out`, or `input` and `output`, would make more sense to me than `in` and `output` as a pair of folders.

File: StreamTemp-Integ.py [hunk 17]
```diff
@@ -68,13 +64,17 @@
...
+
+        forcing_path = os.path.join(os.path.sep, rootDatabase, 'forcing', f_list)  # obs_18basins
+        attr_path = os.path.join(os.path.sep, rootDatabase, 'attr', a_list)
+        if os.path.exists(os.path.join(os.path.sep, rootDatabase, 'in', 'Statistics_basinnorm.json')):
```
Comment (aappling-usgs) [inline StreamTemp-Integ.py]: Joining rootDatabase (which is `[absRoot]/in`) with an additional `in` folder here...is that right? Seems sketchy

File: StreamTemp-Integ.py [hunk 18]
```diff
@@ -98,11 +98,11 @@
         # define training options
         optTrain = default.update(default.optTrainCamels, miniBatch=[BATCH_SIZE, RHO], nEpoch=EPOCH, saveEpoch=saveEPOCH, seed=seed)
         # define output folder for model results
-        exp_name = 'TempDemo'
-        exp_disp = 'FirstRun'
+        exp_name = 'output'
+        #exp_disp = 'demo'
```
Comment (aappling-usgs) [inline StreamTemp-Integ.py:102]: if it's working with the `#`, could fully delete this line now

File: StreamTemp-Integ.py [hunk 19]
```diff
@@ -98,11 +98,11 @@
...
 
 
-        save_path = os.path.join(absRoot, exp_name, exp_disp, \
+        save_path = os.path.join(absRoot, exp_name, \
```
Comment (aappling-usgs) [inline StreamTemp-Integ.py]: how come rootOut doesn't get used in determining the save_path?

File: run_lstm_check_results.ipynb [hunk 20]
```diff
@@ -0,0 +1,739 @@
+{
+ "cells": [
```
Comment (aappling-usgs) [inline run_lstm_check_results.ipynb:2]: I think this file is on the .gitignore list and thus will not be packaged with the rest of the code in the data release. What value do you see in checking it into the github repo?

Comment (wdwatkins) [reply run_lstm_check_results.ipynb:2]: I was working in this notebook while sorting out how to run the model code, and wanted to have something saved while that work was ongoing.  But now that the environment is sorted out and the readme is mostly done it is fairly expendable, so could be removed.

**Comment by wdwatkins** [comment]
Just realized this PR is to Farshid's repo, this should go to `USGS-R:main` right?

File: StreamTemp-Integ.py (hunk 17 above)
Comment (wdwatkins) [reply StreamTemp-Integ.py]: good catch

File: StreamTemp-Integ.py (hunk 19 above)
Comment (wdwatkins) [reply StreamTemp-Integ.py]: Hmm, yeah `exp_name` and `rootOut` are redundant.

File: .gitignore (hunk 1 above)
Comment (wdwatkins) [reply .gitignore]: For git yes, but * isn't handled as expected for copy_zip_repo, so this is still needed it seems.
